minimint-core/src/epoch.rs

```diff
@@ -0,0 +1,227 @@
+use crate::transaction::Transaction;
+use bitcoin_hashes::sha256::Hash as Sha256;
+use hbbft::crypto::{PublicKey, Signature, SignatureShare};
+use hbbft::honey_badger::Batch;
+use itertools::Itertools;
+use minimint_api::encoding::{Decodable, DecodeError, Encodable};
+use minimint_api::{BitcoinHash, FederationModule, PeerId};
+use minimint_derive::UnzipConsensus;
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+
+pub type ConsensusOutcome = Batch<Vec<ConsensusItem>, PeerId>;
+
+#[derive(
+    Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize, UnzipConsensus, Encodable, Decodable,
+)]
+pub enum ConsensusItem {
+    EpochInfo(EpochSignatureShare),
+    Transaction(Transaction),
+    Mint(<minimint_mint::Mint as FederationModule>::ConsensusItem),
+    Wallet(<minimint_wallet::Wallet as FederationModule>::ConsensusItem),
+    LN(<minimint_ln::LightningModule as FederationModule>::ConsensusItem),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub struct EpochSignatureShare(pub SignatureShare);
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub struct EpochSignature(pub Signature);
+
+#[derive(Debug, Clone, Serialize, Deserialize, Encodable, Decodable)]
+pub struct EpochHistory {
+    pub outcome: MerkleOutcome,
+    pub hash: Sha256,
+    pub signature: Option<EpochSignature>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Encodable, Decodable)]
+pub struct MerkleOutcome {
+    pub epoch: u64,
+    pub last_hash: Option<Sha256>,
+    pub items: Vec<(PeerId, Vec<ConsensusItem>)>,
+}
+
+impl MerkleOutcome {
+    pub fn hash(&self) -> Sha256 {
+        let mut bytes: Vec<u8> = vec![];
+        self.consensus_encode(&mut bytes).unwrap();
+        Sha256::hash(&bytes)
+    }
```

Comment:
The name `MerkleOutcome` seems odd since the hashing of the consensus items isn't done as a merkle tree. The referencing of the previous one also doesn't make it a merkle tree. I'm not saying it should be a merkle tree instead of a simple hash, just the name is confusing.

Reply:
I guess it's a hash list which is a subset of a merkle tree, but I can rename.